component/gnap/as/client.go

```diff
@@ -0,0 +1,203 @@
+	"fmt"
+	"io/ioutil"
+	"net/http"
+
+	"github.com/trustbloc/edge-core/pkg/log"
+
+	gnaprest "github.com/trustbloc/auth/pkg/restapi/gnap"
+	"github.com/trustbloc/auth/spi/gnap"
+)
+
+// package as is the Authentication Server client that requests GNAP tokens from the Authorization Server.
+
+//nolint:gochecknoglobals
+var logger = log.New("auth-server-client")
+
+const contentType = "application/json"
+
+// Client requesting Gnap tokens from the Authorization Server.
+type Client struct {
+	signer            gnap.Signer
+	httpClient        *http.Client
+	gnapAuthServerURL string
+}
+
+// NewClient creates a new GNAP authorization client. It requires a signer for HTTP Signature header, an HTTP client
+// and a base URL of the authorization server.
+func NewClient(signer gnap.Signer, httpClient *http.Client, gnapAuthServerURL string) (*Client, error) {
+	if signer == nil {
+		return nil, fmt.Errorf("gnap auth client: missing signer")
+	}
+
+	if httpClient == nil {
+		return nil, fmt.Errorf("gnap auth client: missing http client")
+	}
+
+	if gnapAuthServerURL == "" {
+		return nil, fmt.Errorf("gnap auth client: missing Authorization Server URL")
+	}
+
+	return &Client{
+		signer:            signer,
+		httpClient:        httpClient,
+		gnapAuthServerURL: gnapAuthServerURL,
+	}, nil
+}
+
+// RequestAccess creates a GNAP grant access req then submit it to the server to receive a response with an
+// interact_ref value.
+//nolint:funlen,gocyclo
+func (c *Client) RequestAccess(req *gnap.AuthRequest) (*gnap.AuthResponse, error) {
+	if req == nil {
+		return nil, fmt.Errorf("requestAccess: empty request")
+	}
+
+	mReq, err := json.Marshal(req)
+	if err != nil {
+		return nil, fmt.Errorf("requestAccess: marshal access token error: %w", err)
+	}
+
+	var (
+		sig               []byte
+		marshalledRequest []byte
+	)
+
+	sig, err = c.signer.Sign(mReq)
+	if err != nil {
+		return nil, fmt.Errorf("requestAccess: signature error: %w", err)
+	}
+
+	marshalledRequest, err = json.Marshal(req)
+	if err != nil {
+		return nil, fmt.Errorf("requestAccess: failed to marshal request: %w", err)
+	}
+
+	requestReader := bytes.NewReader(marshalledRequest)
+
+	url := c.gnapAuthServerURL + gnaprest.AuthRequestPath
+
+	httpReq, err := http.NewRequest(http.MethodPost, url, requestReader)
+	if err != nil {
+		return nil, fmt.Errorf("requestAccess: failed to build http request: %w", err)
+	}
+
+	httpReq.Header.Add("Content-Type", contentType)
+	// httpReq.Header.Add("Signature-Input", "TODO") // TODO update signature input
+	httpReq.Header.Add("Signature", base64.URLEncoding.EncodeToString(sig))
+
+	r, err := c.httpClient.Do(httpReq)
+	if err != nil {
+		return nil, fmt.Errorf("requestAccess: failed to post HTTP request to [%s]: %w", gnaprest.AuthRequestPath, err)
+	}
+
+	defer func() {
+		err = r.Body.Close()
+		if err != nil {
+			logger.Warnf("failed to close http request but it has been processed: %w", err)
+		}
+	}()
+
+	if r.StatusCode >= http.StatusBadRequest {
```

Comment:
Probably better to check for success case (known) statusCodes (eg: HTTP 200) and throw error for other cases than this broad failure case validation.

Reply:
anything >= 400 is an error

Reply:
What if Server returns 30x instead of success case of 20x?

Reply:
technically 30x codes are not errors, they indicate a continuation of calls... actually it's a good point for RequestAccess call to return a 30x code so that the user/wallet can expect a followup/redirect call.

Returning 30x is a suggestion, but 40x and 50x are definite HTTP errors

Reply:
For this particular instance, we should only expect HTTP 200 as the success response.

https://www.ietf.org/archive/id/draft-ietf-gnap-core-protocol-09.html#appendix-D.1

```
HTTP/1.1 200 OK
Content-Type: application/json
Cache-Control: no-store

{
    "interact": {
      "redirect":
        "https://server.example.com/interact/4CF492MLVMSW9MKM",
      "finish": "MBDOFXG4Y5CVJCX821LH"
    }
    "continue": {
        "access_token": {
            "value": "80UPRY5NM33OMUKMKSKU"
        },
        "uri": "https://server.example.com/continue"
    },
    "instance_id": "7C7C4AZ9KHRS6X63AJAO"
}
```

Reply:
updated